homeassistant/components/motionmount/entity.py

```diff
@@ -91,6 +95,19 @@ async def _ensure_connected(self) -> bool:
             # The purpose of `_ensure_connected()` is only to make sure we try to
             # reconnect, where failures should not be logged each time
             return False
-        else:
-            _LOGGER.warning("Successfully reconnected to MotionMount")
-            return True
+
+        # Check we're properly authenticated or be able to become so
+        if not self.mm.is_authenticated:
+            if self.pin is None:
+                await self.mm.disconnect()
+                self.config_entry.async_start_reauth(self.hass)
+                return False
+            await self.mm.authenticate(self.pin)
+            if not self.mm.is_authenticated:
+                self.pin = None
+                await self.mm.disconnect()
+                self.config_entry.async_start_reauth(self.hass)
+                return False
```

Comment:
One thing I am wondering is, should we wrap this in an `async with lock:` with an `asyncio.Lock`. Otherwise, every entity will call this code and I can see that going the wrong way, WDYT?

Reply:
I don't think it's necessary. The only place this method is used is in the `async_update` handler of the 'select' entity, as that one is polled.
All other entities become disabled when the connection is lost, so the user can't control them anymore. Hence they do not actively check whether they can bring back the connection, we just wait for the polling loop to come around to revive the connection.

Reply:
I actually think that moving it to the select entity is more clear. I once put it here intended to be called by all entities but it turned out that that was not necessary. It has seen been left here. I'll move it.